skxray/speckle_analysis.py

```diff
@@ -0,0 +1,311 @@
+        intensities of the ROI's of the labeled array according
+        to the pixel list
+        {ROI 1 : intensities of the pixels of ROI 1, ROI 2 : intensities of
+         the pixels of ROI 2}
+    """
+
+    if labels.shape != image.shape:
+        raise ValueError("Shape of the image data should be equal to"
+                         " shape of the labeled array")
+    if index is None:
+        index = np.arange(1, np.max(labels) + 1)
+
+    roi_pix = {n: image[labels == n] for n in index}
+    return roi_pix, index
+
+
+def mean_intensity_sets(images_set, labels):
+    """
+    Mean intensities for ROIS' of the labeled array for different image sets
+
+    Parameters
+    ----------
+    images_set : array
+        images sets
+        shapes is: (len(images_sets), )
+        one images_set is iterable of 2D arrays dimensions are: (rr, cc)
+
+    labels : array
+        labeled array; 0 is background.
+        Each ROI is represented by a distinct label (i.e., integer).
+
+    Returns
+    -------
+    mean_intensity_dict : dict
+        average intensity of each ROI as a dictionary
+        shape len(images_sets)
+        eg: 2 image sets,
+        {image set 1 : len(images in image set 1),
+        image set 2 : len(images in image set 2)}
+
+    index_list : list
+        labels list for each image set
+
+    """
+    mean_intensity_dict = {}
+    index_list = []
+    for n in range(len(images_set)):
+        mean_int, index = mean_intensity(images_set[n], labels)
+        mean_intensity_dict[n] = mean_int
+        index_list.append(index)
+
+    return mean_intensity_dict, index_list
+
+
+def mean_intensity(images, labels, index=None):
+    """
+    Mean intensities for ROIS' of the labeled array for set of images
+
+    Parameters
+    ----------
+    images : array
+        iterable of 2D arrays
+        dimensions are: (rr, cc)
+
+    labels : array
+        labeled array; 0 is background.
+        Each ROI is represented by a distinct label (i.e., integer).
+
+    index : list
+        labels list
+        eg: 5 ROI's
+        index = [1, 2, 3, 4, 5]
+
+    Returns
+    -------
+    mean_intensity : array
+        mean intensity of each ROI for the set of images as an array
+        shape (len(images), number of labels)
+
+    """
+    if labels.shape != images[0].shape[0:]:
+        raise ValueError("Shape of the images should be equal to"
+                         " shape of the label array")
+    if index is None:
+        index = np.arange(1, np.max(labels) + 1)
+
+    mean_intensity = np.zeros((images.shape[0], index.shape[0]))
+    for n, img in enumerate(images):
+        mean_intensity[n] = mean(img, labels, index=index)
+
+    return mean_intensity, index
+
+
+def combine_mean_intensity(mean_int_dict, index_list):
+    """
+    Combine mean intensities of the images(all images sets) for each ROI
+    if the labels list of all the images are same
+
+    Parameters
+    ----------
+    mean_int_dict : dict
```

Comment:
Given that there is a natural order to these values, I think these make more sense as a list than as a dict.  The power of dictionaries is that they give you very fast random access to un-ordered data (or data that if you did order it would be very spare, think reserving space in an array for every valid combination of letters and numbers).  The way that dict is implemented is as a hash table.  The first step is to take your key and reduce it to an integer (hash) which is used to essentially index into an array.  The draw back of this is that the order of the keys is the hash is (nominally) random, thus, when you iterate over the dictionary you can not assume anything about the order.

In this case the application is for looking at the same data set integrated with different intervals which has a very definite ordering.